modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -259,38 +265,136 @@ protected void doStop() {
+                return next.process();
+            }
+            RequestMetrics metrics = (RequestMetrics) metricsData.get();
+            metrics.requestCount++;
+            return next.process().doOnError(throwable -> {
+                logger.debug("Detected error in RetryMetricsTracker", throwable);
+                metrics.errorCount++;
+            }).doOnSuccess(response -> {
+                if (isErrorCode(response.getStatusCode())) {
+                    metrics.errorCount++;
+                    if (response.getStatusCode() == RestStatus.TOO_MANY_REQUESTS.getStatus()) {
+                        metrics.throttleCount++;
+                    }
+                }
+            });
+        }
+
+        @Override
+        public HttpPipelinePosition getPipelinePosition() {
+            return HttpPipelinePosition.PER_RETRY;
+        }
+
+        private static boolean isErrorCode(int statusCode) {
+            return statusCode < 200 || statusCode > 299;
+        }
+    }
+
+    private static final class RequestMetricsTracker implements HttpPipelinePolicy {
+        private static final String ES_REQUEST_METRICS_CONTEXT_KEY = "_es_azure_repo_request_stats";
+        private static final Logger logger = LogManager.getLogger(RequestMetricsTracker.class);
         private final OperationPurpose purpose;
-        private final SuccessfulRequestHandler onSuccessfulRequest;
+        private final RequestMetricsHandler requestMetricsHandler;
 
-        private SuccessfulRequestTracker(OperationPurpose purpose, SuccessfulRequestHandler onSuccessfulRequest) {
+        private RequestMetricsTracker(OperationPurpose purpose, RequestMetricsHandler requestMetricsHandler) {
             this.purpose = purpose;
-            this.onSuccessfulRequest = onSuccessfulRequest;
+            this.requestMetricsHandler = requestMetricsHandler;
         }
 
         @Override
         public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
-            return next.process().doOnSuccess(httpResponse -> trackSuccessfulRequest(context.getHttpRequest(), httpResponse));
+            final RequestMetrics requestMetrics = new RequestMetrics();
+            final long requestStartTimeInMillis = System.currentTimeMillis();
+            context.setData(ES_REQUEST_METRICS_CONTEXT_KEY, requestMetrics);
+            return next.process().doOnSuccess((httpResponse) -> {
+                requestMetrics.statusCode = httpResponse.getStatusCode();
+                requestMetrics.timeToResponseInMillis = System.currentTimeMillis() - requestStartTimeInMillis;
+                trackCompletedRequest(context.getHttpRequest(), requestMetrics);
+            }).doOnError(throwable -> {
+                logger.debug("Detected error in RequestMetricsTracker", throwable);
+                trackCompletedRequest(context.getHttpRequest(), requestMetrics);
```

Comment:
One thing that _might_ manifest in this way is a request timeout, which we probably want to see reflected in the metrics. Basically any time you get a response back from the server, even 5xx responses, we execute `doOnSuccess`. "Success" here just means a request with a response, it's a successful `Mono` not a successful HTTP request. The errors occur when things happen like the connection is dropped.

Reply:
It appears as though S3 client records `HttpRequestTime` even for requests that end in an exception (assuming timeouts, connection failures etc.), see [here](https://github.com/aws/aws-sdk-java/blob/864baf847e35b6c987ad08d81738050d329c515c/aws-java-sdk-core/src/main/java/com/amazonaws/http/AmazonHttpClient.java#L1369-L1403).

The time we record appears to be all of these durations added together, so that would exclude the time taken to pause between retries. I'll make the Azure client consistent with that.